### ISSUE TITLE
Implement reCaptcha for new account creation

### ISSUE THREAD

**Issue by davidbau** [PencilCode/pencilcode#165]
jsperf.com has been flooded with spammers hitting his site, and so before that happens here, I'd like us to implement reCaptcha for account creation (and maybe at some point a flag where we can enable it for logins-every-n-days).

Here is information on the reCaptcha 2.0 API - looks relatively straightforward, but it's a project.

https://developers.google.com/recaptcha/intro

**Comment by weihang7** [comment]
Implementing it is fairly straightforward, but what should be done in countries like China where reCaptcha is not accessible?

**Comment by davidbau** [comment]
Pencil Code is hosted on Google Compute Engine.  Is it accessible in China?

David

**Comment by xinan** [comment]
Just asked one of my friends in China to test. It is accessible in China.

**Comment by davidbau** [comment]
This stack overflow question suggests a local alternative.  http://stackoverflow.com/questions/23780387/recaptcha-availability-in-china

We might investigate its use for China-only.  Not sure how it works:
http://www.yinxiangma.com/

**Comment by davidbau** [comment]
Thinking about the spam scenario....

Normally requiring "reCaptcha for every new account and every password change" makes sense.  But there should be a defensive setting to allow us to "require a reCaptcha on every save".  That would allow us to keep the service open if targeted by spammers.

**Comment by weihang7** [comment]
The server side verification of reCaptcha involves a secret token. Should I put that in this public repository as well?

**Comment by davidbau** [comment]
No.  You can work with me to set up a secret token; write code to load it
in from a file; and I can put it on a secrets directory in the server.

**Comment by weihang7** [comment]
Unfortunately I wasn't able to finish implementing it before going back to China. I will be back in the US in September. If anyone else would like to work on it a WIP is at pencilcode/recaptcha.

**Comment by davidbau** [comment]
Thanks Frank!  Enjoy summer break in China!
